src/dymoprint/lib/render_engines/margins.py

```diff
@@ -0,0 +1,115 @@
+        min_width_px: float = 0,
+    ):
+        super().__init__()
+        labeler_horizontal_margin_px, labeler_vertical_margin_px = labeler_margin_px
+        assert visible_horizontal_margin_px >= 0
+        assert labeler_horizontal_margin_px >= 0
+        assert labeler_vertical_margin_px >= 0
+        assert not max_width_px or max_width_px >= 0
+        assert min_width_px >= 0
+        self.mode = mode
+        self.justify = justify
+        self.visible_horizontal_margin_px = visible_horizontal_margin_px
+        self.labeler_horizontal_margin_px = labeler_horizontal_margin_px
+        self.labeler_vertical_margin_px = labeler_vertical_margin_px
+        self.max_width_px = max_width_px
+        self.min_width_px = min_width_px
+        self.render_engine = render_engine
+
+    def calculate_visible_width(self, payload_width_px: int) -> float:
+        minimal_label_width_px = (
+            payload_width_px + self.visible_horizontal_margin_px * 2
+        )
+        if self.max_width_px is not None and minimal_label_width_px > self.max_width_px:
+            raise BitmapTooBigError(minimal_label_width_px, self.max_width_px)
+
+        if self.min_width_px > minimal_label_width_px:
+            label_width_px = self.min_width_px
+        else:
+            label_width_px = minimal_label_width_px
+        return label_width_px
+
+    def render(self, context: RenderContext) -> tuple[Image.Image, dict[str, float]]:
+        payload_bitmap = self.render_engine.render(context)
+        payload_width_px = payload_bitmap.width
+        label_width_px = self.calculate_visible_width(payload_width_px)
+        padding_px = label_width_px - payload_width_px  # sum of margins from both sides
+
+        if self.justify == "left":
+            horizontal_offset_px = self.visible_horizontal_margin_px
+        elif self.justify == "center":
+            horizontal_offset_px = padding_px / 2
+        elif self.justify == "right":
+            horizontal_offset_px = padding_px - self.visible_horizontal_margin_px
+        assert horizontal_offset_px >= self.visible_horizontal_margin_px
+
+        # In print mode:
+        # ==============
+        # There is a gap between the printer head and the cutter (for the sake of this
+        # example, let us say it is DX pixels wide).
+        # We assume the printing starts when the print head is in offset DX from the
+        # label's edge (just under the cutter).
+        # After we print the payload, we need to offset the label DX pixels, in order
+        # to move the edge of the printed payload past the cutter, otherwise the cutter
+        # will cut inside the printed payload.
+        # Afterwards, we need to offset another DX pixels, so that the cut will have
+        # some margin from the payload edge. The reason we move DX pixels this time, is
+        # in order to have symmetry with the initial margin between label edge and start
+        # of printed payload.
```

Comment:
This offset business is so confusing. I'm trying to make it a bit more clear:

```suggestion
        # Subsequently, in order to achieve symmetry so that the final margin matches
        # the initial margin, we add another offset of DX pixels, so that the cut
        # will have that same margin from both sides of the payload edge. In summary,
        # due to the gap between the printer head and the cutter of DX pixels,
        # printing an initial margin of 0 pixels and a final margin of 2 * DX pixels
        # leaves an effective margin of DX pixels on both sides.
```

Reply:
And this?

Reply:
Perfect. Thanks.